File: core/src/main/java/org/mifos/mobilewallet/core/domain/usecase/EditCard.java
package org.mifos.mobilewallet.core.domain.usecase;

import android.util.Log;

import org.mifos.mobilewallet.core.base.UseCase;
import org.mifos.mobilewallet.core.data.fineract.api.GenericResponse;
import org.mifos.mobilewallet.core.data.fineract.repository.FineractRepository;
import org.mifos.mobilewallet.core.domain.model.Card;

import javax.inject.Inject;

import rx.Subscriber;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by ankur on 22/May/2018
 */

public class EditCard extends UseCase<EditCard.RequestValues, EditCard.ResponseValue> {

    private final FineractRepository mFineractRepository;

    @Inject
    public EditCard(FineractRepository fineractRepository) {
        mFineractRepository = fineractRepository;
    }

    @Override
    protected void executeUseCase(RequestValues requestValues) {

        mFineractRepository.editSavedCard(requestValues.cardId, requestValues.card)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(new Subscriber<GenericResponse>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {

                    }

                    @Override
                    public void onNext(GenericResponse genericResponse) {
                        Log.d("qxz", "onNext: card updated");
                    }
                });

    }

    public static final class RequestValues implements UseCase.RequestValues {

        private final int cardId;
        private final Card card;

        public RequestValues(int cardId, Card card) {
            this.cardId = cardId;
            this.card = card;
        }
    }

    public static final class ResponseValue implements UseCase.ResponseValue {

    }
}

File: core/data/src/main/java/org/mifospay/core/data/domain/usecase/savedcards/EditCard.kt
package org.mifospay.core.data.domain.usecase.savedcards

import org.mifospay.core.data.base.UseCase
import com.mifospay.core.model.entity.savedcards.Card
import org.mifospay.core.data.fineract.repository.FineractRepository
import org.mifospay.core.network.GenericResponse
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import javax.inject.Inject

/**
 * Created by ankur on 22/May/2018
 */
class EditCard @Inject constructor(private val mFineractRepository: FineractRepository) :
    UseCase<EditCard.RequestValues, EditCard.ResponseValue>() {

    class RequestValues(val clientId: Int, val card: Card) : UseCase.RequestValues
    class ResponseValue : UseCase.ResponseValue

    override fun executeUseCase(requestValues: RequestValues) {
        mFineractRepository.editSavedCard(requestValues.clientId, requestValues.card)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeOn(Schedulers.io())
            .subscribe(object : Subscriber<GenericResponse?>() {
                override fun onCompleted() {}
                override fun onError(e: Throwable) {
                    useCaseCallback.onError(e.toString())
                }

                override fun onNext(t: GenericResponse?) {
                    useCaseCallback.onSuccess(ResponseValue())
                }
            })

    }
}
